metabase/metabase
# ESLint for detecting the use of Metabase links

Milestone 2 of #36966

### Description

Tested with `time yarn lint-eslint-pure` both `master` and this PR yielded similar results over a couple of tries ~130s on my M1 Macbook pro.

Using `TIME=all yarn lint-eslint-pure` will show that this rule only took ~30ms and it accounts for 0.0% of the total time linting

Rule                                                   | Time (ms) | Relative
:------------------------------------------------------|----------:|--------:
no-unconditional-metabase-links-render                 |    33.337 |     0.0%

This adds a rule to detect that when dealing with Metabase links we should either
1. conditionally render the links or
2. disable the rule with a reason

So we don't shoot ourselves in the foot in the future after this feature has been fully shipped. Linting is necessary because we will definitely forget about this feature in no time, and this will prevent the mishandling with Metabase links in the future.

#### Types of links that this rule detects
 1. `MetabaseSettings.learnUrl(string)`
 2. `MetabaseSettings.docsUrl(string)`
 3. `getDocsUrl` selector from "metabase/selectors/settings"
 4. `getLearnUrl` selector from "metabase/selectors/settings"
 5. inline string "metabase.com/docs/"
 6. inline string "metabase.com/learn/"

### Changes
1. The ESLint rule
2. The rule's test
3. Other files are the result of ESLint working, and those are places we don't want to conditionally render the link with the specified reason.

### How to verify

1. Go to any existing `.jsx` or `.tsx` file, or even create a new one if you may.
4. Try to add any type of Metabase links from above
5. You should see the ESLint error with a message telling you how to fix it.

### Demo

- ![image](https://github.com/metabase/metabase/assets/1937582/3b8dd9f7-c5f1-43dc-8cd0-fc702a91c382)
- ![image](https://github.com/metabase/metabase/assets/1937582/2673e60b-51a6-428f-a081-d7c59ef983e5)
- ![image](https://github.com/metabase/metabase/assets/1937582/4b4dfc6b-7e25-472e-a6a2-4e52e4a9fbd3)
- ![image](https://github.com/metabase/metabase/assets/1937582/930ec0f9-1b44-414b-a42f-3aa7e8e6bdef)
- ![image](https://github.com/metabase/metabase/assets/1937582/bb144902-8909-48ef-9ad0-400476e6543a)
- ![image](https://github.com/metabase/metabase/assets/1937582/aad0e1f0-990a-4dc9-90f4-892703b05c65)
- ![image](https://github.com/metabase/metabase/assets/1937582/48586bc3-df07-404f-9835-e54532b8c529)
- ![image](https://github.com/metabase/metabase/assets/1937582/cc32ab9d-2255-424b-bbb1-b4dc41b7cf26)



### Checklist
